pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,192 @@
+from typing import Union, Optional, Callable, Any, List
+import torch
+from torch.utils.data import DataLoader
+from torch import nn
+from torch.optim.optimizer import Optimizer
+from torch.optim.lr_scheduler import ExponentialLR
+import pystiche
+from pystiche.optim import (
+    OptimLogger,
+    default_transformer_epoch_optim_loop,
+)
+from ..common_utils import batch_up_image
+from .modules import ulyanov_et_al_2016_transformer
+from .loss import (
+    UlyanovEtAl2016PerceptualLoss,
+    ulyanov_et_al_2016_perceptual_loss,
+)
+
+from .data import (
+    ulyanov_et_al_2016_style_transform,
+    ulyanov_et_al_2016_content_transform,
+    ulyanov_et_al_2016_dataset,
+    ulyanov_et_al_2016_image_loader,
+    ulyanov_et_al_2016_images,
+)
+from .utils import (
+    ulyanov_et_al_2016_preprocessor,
+    ulyanov_et_al_2016_postprocessor,
+    ulyanov_et_al_2016_optimizer,
+)
+
+
+__all__ = [
+    "ulyanov_et_al_2016_transformer",
+    "ulyanov_et_al_2016_perceptual_loss",
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_image_loader",
+    "ulyanov_et_al_2016_training",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_stylization",
+]
+
+
+class DelayedExponentialLR(ExponentialLR):
+    def __init__(
+        self, optimizer: Optimizer, gamma: float, delay: int, **kwargs: Any
+    ) -> None:
+        self.delay = delay
+        super().__init__(optimizer, gamma, **kwargs)
+
+    def get_lr(self) -> List[float]:
+        exp = self.last_epoch - self.delay + 1
+        if exp > 0:
+            return [base_lr * self.gamma ** exp for base_lr in self.base_lrs]
+        else:
+            return self.base_lrs
+
+
+def ulyanov_et_al_2018_lr_scheduler(
+    optimizer: Optional[Optimizer] = None, impl_params: bool = True,
```

Comment:
But for the `optimizer` the `transformer` is needed. For this reason, if the `optimizer` is not known here, `none` should be returned and this should be handled later when the `transformer` is known. For example in the `default_transformer_epoch_optim_loop` as it is done there.

Reply:
1. You cannot create an `LRScheduler` without an `optimizer`. Try calling `ulyanov_et_al_2018_lr_scheduler()` without parameters and this will raise an `Exception`.
2. You are right that an `LRScheduler` ultimately requires the `transformer` first, but I don't see how this is a problem. If I'm missing something, could you describe an scenario where this is a problem?
3. Just saw that: it should be `2016` instead of `2018` in the signature.